core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -1175,6 +1237,21 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
     deleted
   }
+
+  /** NOTE: 'task' should ensure it executes 'endProcessing' at the end */
```

Comment:
Hmm, perhaps this method could take care of calling `endProcessing` too (e.g. by wrapping the task)? Should be just a small adjustment at the call site.

Reply:
`finally` statement in `mergeApplicationListing` makes it be complicated, because we should handle the reentrance of lock.

If we move endProcessing from `finally` statement in `mergeApplicationListing` to the end of task here, `processing` on the compaction task would be executed earlier than calling `endProcessing` on the listing task. Marking lock from compaction task would succeed but effectively no-op, and releasing lock from listing task would remove the mark for compaction task as well, which makes the compaction task run without proper lock.

So either we need to make lock much smarter, or document the requirement on caller side. I'm feeling that former one is more complicated than latter one.